### PR TITLE
Maintain minimum number of new lines when formatting methods in a class/service

### LINKED ISSUE #41462: [Improvement]: Preserve a minimum number of blank lines when formatting module-level defintions
### Description

Refer to the below code, which doesn't get changed as per our current formatting behaviour. 
```ballerina
# function foo
# 
function foo() returns int {
    return 1;
}
# function bar
# 
function bar() returns int {
    return 1;
}
# function baz
# 
function baz() returns int {
    return 1;
}
```

Due to this behaviour, all the generated client connectors from the openAPI tool contains the same style, which makes it less readable IMO. (E.g: https://github.com/ballerina-platform/openapi-connectors/blob/main/openapi/stripe/client.bal)

Therefore, it would be a better approach to insert/preserve at-least a single blank line in-between the function definitions, as the default formatting behaviour.  

### Describe your problem(s)

_No response_

### Describe your solution(s)

_No response_

### Related area

-> Formatter

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by poorna2152**
Fixes #41462 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by NipunaRanasinghe — COMMENTED** [review]
It seems these minimum/maximum line numbers are not configurable via `FormattingOptions`. Is that expected? Are there any plans to support that in the future?

File: misc/formatter/modules/formatter-core/src/test/resources/declarations/class-definition/source/class_definition_declaration_18.bal [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+        self.name = name;
+        self.age = age;
+    }
+    public function getName() returns string {
+        return self.name;
+    }
+    public function getAge() returns int {
+        return self.age;
+    }
+}
```
Comment (NipunaRanasinghe) [inline misc/formatter/modules/formatter-core/src/test/resources/declarations/class-definition/source/class_definition_declaration_18.bal:15]: Do we have a test case to check what happens if there are more than one blank lines? (if not, probably we can use this same test resource to verify the output in that scenario as well)

**Comment by poorna2152** [comment]
> It seems these minimum/maximum line numbers are not configurable via `FormattingOptions`. Is that expected? Are there any plans to support that in the future?

We can add a formatting option here to configure the minimum number newline after method definitions. This also be added in the [formatting configuration feature](https://github.com/ballerina-platform/ballerina-lang/issues/41280). WDYT

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java [hunk 2]
```diff
@@ -3886,6 +3904,17 @@ private <T extends Node> boolean isMultilineModuleMember(T node) {
         }
     }
 
+    private <T extends Node> boolean isMemberOfScope(T node) {
+        switch (node.kind()) {
+            case OBJECT_METHOD_DEFINITION:
```
Comment (LakshanWeerasinghe) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java:3909]: What is the syntax kind we are getting for remote method definition? Add a test case to cover remote methods as well.

Comment (poorna2152) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java:3909]: In the case of remote methods the syntax kind is `RESOURCE_ACCESSOR_DEFINITION`. Will add a test case for that

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java [hunk 3]
```diff
@@ -3886,6 +3904,17 @@ private <T extends Node> boolean isMultilineModuleMember(T node) {
...
+            case RESOURCE_ACCESSOR_DEFINITION:
+            case OBJECT_FIELD:
+                return true;
+            default:
+                return false;
+        }
+    }
```
Comment (dulajdilshan) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java:3916]: The private function's name: `isMemberOfScope` and its implementation are quite mismatched to me. What I see here is, that we are checking the node kind across a set of nodes and returning a true/false. What is the intention of this check?

Comment (poorna2152) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java:3916]: This issue is related to functions defined within a scope(service classes, classes, services). Thus here given a Node it identifies whether is node is in a scope.

Comment (dulajdilshan) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java:3916]: Got it!

**Review by dulajdilshan — DISMISSED** [review]

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java [hunk 4]
```diff
@@ -311,8 +312,25 @@ public FunctionDefinitionNode transform(FunctionDefinitionNode functionDefinitio
         }
         NodeList<Node> relativeResourcePath = formatNodeList(functionDefinitionNode.relativeResourcePath(), 0, 0, 0, 0);
         FunctionSignatureNode functionSignatureNode = formatNode(functionDefinitionNode.functionSignature(), 1, 0);
+        int trailingNL = env.trailingNL;
+        if (isMemberOfScope(functionDefinitionNode)) {
+            ChildNodeList parentChildren = functionDefinitionNode.parent().children();
```
Comment (lochana-chathura) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java:317]: This is not the correct place to handle `itemTrailingNL` for service and class members. I think we need to handle them at the parent level itself in [1] and [2]. I expect the implementation to be similar to `formatModuleMembers()` we have implemented in [3] for formatting the module members.

[1]. https://github.com/ballerina-platform/ballerina-lang/blob/2752ae1d5f93a9f99684812125f0530dd614ebe7/misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java#L3465
[2]. https://github.com/ballerina-platform/ballerina-lang/blob/2752ae1d5f93a9f99684812125f0530dd614ebe7/misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java#L703
[3]. https://github.com/ballerina-platform/ballerina-lang/blob/2752ae1d5f93a9f99684812125f0530dd614ebe7/misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java#L291

**Review by lochana-chathura — CHANGES_REQUESTED** [review]

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java [hunk 5]
```diff
@@ -3886,6 +3904,17 @@ private <T extends Node> boolean isMultilineModuleMember(T node) {
         }
     }
 
+    private <T extends Node> boolean isMemberOfScope(T node) {
+        switch (node.kind()) {
+            case OBJECT_METHOD_DEFINITION:
+            case RESOURCE_ACCESSOR_DEFINITION:
+            case OBJECT_FIELD:
```
Comment (lochana-chathura) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java:3911]: Need to include the kind for the type-inclusion which is `TYPE_REFERENCE` here as well.

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java (hunk 4 above)
Comment (poorna2152) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormattingTreeModifier.java]: Fixed

**Review by NipunaRanasinghe — APPROVED** [review]

**Comment by poorna2152** [comment]
Merged to a feature branch due to tests failing in the `openapi` and `graphql-cli`.

**Comment by poorna2152** [comment]
Sending to master in #42370

**Pull request merged**